src/goto-programs/destructor_tree.cpp

```diff
@@ -0,0 +1,108 @@
+#include "destructor_tree.h"
+
+/// Adds a destructor to the current stack, either attaching itself to the
+/// closest previous node, or an appropriate rewind target.
+void destructor_treet::add(const codet &destructor)
+{
+  // If we have a stored rewind target, attach node to that then wipe rewind.
+  auto previous_node = get_current_node();
+  auto new_node = destruction_graph.add_node(destructor);
+  if(rewind_target)
+  {
+    auto rewind_target_index = *rewind_target;
+    rewind_target = {};
+    if(!destruction_graph.has_edge(rewind_target_index, new_node))
+      destruction_graph.add_edge(rewind_target_index, new_node);
+  }
+  else
+  {
+    destruction_graph.add_edge(previous_node, new_node);
+  }
+}
+
+/// Fetches the destructor value for the passed-in node identity.
+optionalt<codet> &destructor_treet::get(std::size_t identity)
+{
+  return destruction_graph[identity].destructor_value;
+}
+
+/// Finds the node two branches intersect on and then returns it.
+/// This should be used when you want to find out what parts of the two
+/// branches are common to each other.
+const intersection_resultt destructor_treet::get_intersection_node(
+  optionalt<std::size_t> left_optional,
+  optionalt<std::size_t> right_optional)
+{
+  if(destruction_graph.empty() || !right_optional || !left_optional)
+    return {0};
+
+  std::size_t right_identity = *right_optional, left_identity = *left_optional,
+              left_unique_count = 0, right_unique_count = 0;
+  while(right_identity != left_identity)
+  {
+    if(right_identity > left_identity)
+      right_identity = destruction_graph[right_identity].in.begin()->first,
+      right_unique_count++;
+    else
+      left_identity = destruction_graph[left_identity].in.begin()->first,
+      left_unique_count++;
+  }
+
+  return {right_identity, left_unique_count, right_unique_count};
+}
+
+/// Builds a destructor stack starting at the passed-in identity.
+/// An identity is the internal ID that you get that relates to the
+/// position in the psudeo-tree of the node that you want.
+/// \param end_identity The identity to signal a stop to destructing.
+///     List won't include this node. If empty, root of the current stack.
+/// \param starting_identity The identity to start destructing from. if empty, top of the current stack.
+/// \return collection of destructors that should be called for the range supplied.
+const std::vector<codet> destructor_treet::get_destructors(
+  optionalt<std::size_t> end_identity,
+  optionalt<std::size_t> starting_identity)
+{
+  if(destruction_graph.empty())
+    return {};
+
+  std::size_t next_id =
+    starting_identity ? *starting_identity : get_current_node();
+  std::size_t end_id = end_identity ? *end_identity : 0;
```

Comment:
You could use `std::size_t end_id = end_identity.value_or(0);` and similarly for the above.